### PR TITLE
Adhere to Base.copy!(x, i, y, j, n) behaviour for n <= 0

### PR DESCRIPTION
**Pull request by alyst**
This is consistent with Base `copy!(x, i, y, j, 0)`: no checks for `i` or `j` are done.
IRL it's observed during `join(df1, df2, kind=:inner)` when the result contains no rows.

### CODE REVIEW

**Comment by alyst** [comment]
The usual `Date` CI failures seem unrelated.

**Comment by nalimilan** [comment]
Why not, if that's consistent with Base. Can you put the tests in a testset and make comments more explicit about the fact that we test that no error is thrown for out of bounds indices?

**Comment by alyst** [comment]
added testsets

**Comment by alyst** [comment]
Sorry for pushing another commit after the approval, but it's related, so I though it should go in the same PR.
It actually improves the conformance to Base `copy!()`: the negative length (`n`) should be checked before the indices checks, but not in the case of the default `n`.
Otherwise `BoundsError` may confuse the user by referencing the copy range that the user never specified.

File: src/array.jl [hunk 1]
```diff
@@ -396,14 +396,21 @@ copy(A::CategoricalArray) = deepcopy(A)
 CatArrOrSub{T, N} = Union{CategoricalArray{T, N},
                           SubArray{<:Any, N, <:CategoricalArray{T}}} where {T, N}
 
+function copy!(dest::CatArrOrSub, dstart::Integer,
```
Comment (nalimilan) [inline src/array.jl]: AFAICT we don't need to define this method, we can rely on the `AbstractArray` fallback. Just remove the default value of `n` in the method below.

**Comment by alyst** [comment]
CI passes on 0.6

**Pull request merged**